NGitLab.Mock/Clients/ClusterClient.cs

```diff
@@ -14,6 +14,12 @@ public ClusterClient(ClientContext context, int projectId)
             _projectId = projectId;
         }
 
+        public ClusterClient(ClientContext context, string projectId)
```

Comment:
We indeed don't have the choice to create a new constructor everywhere for adding the support for project/group paths.

However, instead of just a string, it might be better to use a dedicated struct as a parameter that would contains either a long for the id (that would help for https://github.com/ubisoft/NGitLab/issues/491) or a string for the path.

```csharp
public readonly struct ProjectIdOrPath
{
    private readonly long _id;
    private readonly string? _path;

    public ProjectIdOrPath(long id) => _id = id;
    public ProjectIdOrPath(string path) => _path = path;

    public static implicit operator ProjectIdOrPath(long id) => new(id);
    public static implicit operator ProjectIdOrPath(string path) => new(path);

    internal readonly string ValueAsString => _path ?? _id.ToString(CultureInfo.InvariantCulture);
    internal readonly string ValueAsUriParameter => Uri.EscapeDataString(_path ?? _id.ToString(CultureInfo.InvariantCulture));

    public override string ToString() => ValueAsString;
}
```

- The new constructors will use `ProjectIdOrPath`. Implicit operators will make it simple to use.
- The old one should still work and at some point, can be marked as obsolete
- Wherever it's needed, we could use `ValueAsUriParameter` to construct URLs, to avoid doing that by hand every time
- We can have a similar `GroupIdOrPath`

Reply:
Note that you can add additional implicit converters if it makes the code easier to use. For instance,

````
public static implicit operator ProjectIdOrPath(Project project) => new(project.Id);
````

Reply:
Awesome! It's a great design, I will follow it later.

Reply:
There is totally no difference between `ProjectIdOrPath` and `GroupIdOrPath`.

I think we should use the same struct, or make a base struct at least.

Reply:
While there is no difference yet, it can change in the future. As I mentioned, you could add specific implicit converters to the struct.

Also, having a struct per concept allows to clearly see where a concept is used. It can also improve readability and avoid issues when a method accepts multiple similar parameters. For instance, `Get(Id groupId, Id projectId)` vs `Get(GroupId group, ProjectId project)`. I don't think there is a method with such signature yet, but it's more future proof.

If there are many similar structs in the future, we may use a code generator to write the boilerplate. I don't think we need this complexity at the moment as there are only 2 structs.

> make a base struct at least

Structs cannot be inherited

Reply:
I think `groupId` will never be passed with `projectId` at a single request.

`projectId` is a unique identifier of a specific project. It contains the groupId internally if it's a namespaced path.

And in GitLab docs, there's a [**single** section describing namespaced path](https://docs.gitlab.com/ee/api/rest/index.html#namespaced-path-encoding). 
Both `projectId` and `groupId` are pointed to this section.

> Structs cannot be inherited

Can we use `record` or just `class`?

Reply:
So, to sum up. I made a struct named `IdOrNamespacedPath` for both group and project for now.

(It's called `NamespacedPath` in GitLab docs, but `PathWithNamespace` may be clearer. What are your thoughts on this?)

Reply:
Finally, I made 2 records type `ProjectId` and `GroupId` that inherits from `record IdOrNamespacedPath`.

Reply:
- I would tend to say that struct is still a valid option, even if it's duplicated. The project at the moment don't make an heavy use of either struct or records.
- I would have thought that the project already has nullable enabled ^^. Though it doesn't seem to be the case.

@louis-z , @meziantou ,
- Is the use of record in this case a valid option? Should we enforce a choice on tiny value types as those IDs?
- I suppose that not having nullable enabled on project might be because of the amount of work to do so. Though, is having `#nullable enable` acceptable on new files?

Reply:
> Is the use of record in this case a valid option? Should we enforce a choice on tiny value types as those IDs?

You'll need to handle null values everywhere. If the goal is to share some behavior, you can also use an interface (IIdOrPathAddressable).

> I suppose that not having nullable enabled on project might be because of the amount of work to do so. Though, is having #nullable enable acceptable on new files?

Historic reasons, but it would be nice to use Nullable Reference Types at least on new code.

Reply:
@Toa741 @meziantou I met a problem while use `#nullable enable` for a single file:

```
RS0037: PublicAPI.txt is missing '#nullable enable', so the nullability annotations of API isn't recorded. It is recommended to enable this tracking.
```

If I put `#nullable enable` to `PublicAPI.txt`, all other API without Nullable Reference will be broken:

```
RS0036: Symbol '***' is missing nullability annotations in the declared API
RS0041: Symbol '***' uses some oblivious reference types
```